CORDA-3777 Reload after every checkpoint

lankydan opened the pull request:
Enable reloading of a flow after every checkpoint is saved. This
includes reloading the checkpoint from the database and recreating the
fiber.

When a flow and its `StateMachineState` is created it checks the node's
config to see if the `reloadCheckpointAfterSuspend` is set to true. If it is
it initialises `StateMachineState.reloadCheckpointAfterSuspendCount`
with the value 0. Otherwise, it remains `null`.

This count represents how many times the flow has reloaded from its
checkpoint (not the same as retrying). It is incremented every time the
flow is reloaded.

When a flow suspends, it processes the suspend event like usual, but
it will now also check if `reloadCheckpointAfterSuspendCount` is not
`null` (that it is activated) and process a 
`ReloadFlowFromCheckpointAfterSuspend`event, if and only if 
`reloadCheckpointAfterSuspendCount` is greater than
`CheckpointState.numberOfSuspends`. The retry event handles the
reloading.

This means idempotent flows can reload from the start and not reload
again until reaching a new suspension point.

Flows that skip checkpoints can reload from a previously saved
checkpoint (or from the initial checkpoint) and will continue reloading
on reaching the next new suspension point (not the suspension point that
it skipped saving).

A new staff member has been added to the flow hospital to force it to
ignore the `ReloadFlowFromCheckpointException` thrown if the flow fails
deserialization when reloading.

lankydan committed:
CORDA-3777 Reload after every checkpoint

Enable reloading of a flow after every checkpoint is saved. This
includes reloading the checkpoint from the database and recreating the
fiber.

`reloadCheckpointAfterSuspend` is a field added to `Event.Suspend`. When
creating the event it checks the configuration to determine if it should
reload after processing the suspend event.

If it should reload, then it processes an `Event.RetryFlowFromSafePoint`
straight after completing the suspend. This handles the reloading.

This works because the flag on the suspend event, sets another flag on
the `StateMachineState`. When recreating the transient state
(`StateMachineState`), the same flag is set to false. This means when
the fiber is restarted (it resumes at the point after completing the
suspend event in `FlowStateMachineImpl.suspend`), it does not retry
again.

Without this switching between the flags the flow would be stuck in an
infinite loop of retrying.

A new staff member has been added to the flow hospital to force it to
ignore the `ReloadFlowFromCheckpointException` thrown if the flow fails
deserialization when reloading.

rick-r3 commented:
_Looks good to me.  Happy for @williamvigorr3 etc to review further_.  **<---- forget this.  Changed my mind.  See my review comment below.**

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -488,10 +493,11 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
+                    maySkipCheckpoint = skipPersistingCheckpoint,
+                    fiber = this.checkpointSerialize(context = serializationContext.value),
+                    progressStep = logic.progressTracker?.currentStep,
+                    reloadCheckpointAfterSuspend = serviceHub.value.configuration.reloadCheckpointAfterSuspend
```
Mmmm.... maybe it's not quite this easy!

e.g. imagine this is an `IdempotentFlow` that suspends many time (you could probably build a test one).  There will only be one checkpoint that we keep restoring from, but we do also want to test retrying at each suspension point after the initial checkpoint in case some transient state in the flows (statics etc) breaks something.

Now I'm thinking the idea of having the hospital track which suspension points have been retried from (via medical history records) is a better idea, and then you can check whether the flow already retried at a particular suspension point.

Or does the code already work that way?  My impression, without detailed study, is that it just goes to the next suspension after a retry and thus wouldn't work, but happy to be proven wrong.

lankydan committed:
CORDA-3777 Use count instead of flag

To make the reloading work for idempotent flows and for flows that skip
saving their checkpoints, a count is maintained for what suspension
point was reached and reloaded from.

This means idempotent flows can reload from the start and not reload
again until reaching a new suspension point.

Flows that skip checkpoints can reload from a previously saved
checkpoint (or from the initial checkpoint) and will continue reloading
on reaching the next new suspension point (not the suspension point that
it skipped saving).

lankydan committed:
CORDA-3777 Reload test

lankydan committed:
CORDA-3777 Fix after rebase

lankydan committed:
CORDA-3777 Remove redundant code

lankydan committed:
CORDA-3777 Enable via system property

lankydan committed:
CORDA-3777 Detekt and make the retry test better

lankydan committed:
CORDA-3777 Detekt in flow creator

lankydan committed:
CORDA-3777 Set mock config value in mock network tests

lankydan committed:
CORDA-3777 Fix unused variables

lankydan committed:
CORDA-3777 Rename nodes in tests

lankydan committed:
CORDA-3777 Start nodes in parallel

lankydan committed:
CORDA-3777 Tidy up after rebase

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StaffedFlowHospital.kt:
```diff
@@ -665,6 +666,25 @@ class StaffedFlowHospital(private val flowMessaging: FlowMessaging,
+            history: FlowMedicalHistory
+        ): Diagnosis {
+            return if (newError.mentionsThrowable(ReloadFlowFromCheckpointException::class.java)) {
+                Diagnosis.TERMINAL
```
Should we retain the checkpoint for analysis ?  Does that suggest observation (instead)?  Maybe that makes your testing harder.

rick-r3 replied:
Maybe a test mode (for you) where it throws, but normally keeps in for observation.

lankydan replied:
Mmm I think if we are going to do that, we should both throw the exception and hospitalize the flow. Will be more work but I think that is the best idea.

I agree that keeping the checkpoint does make sense. Maybe this should hook into the client id work? In that, we plan to keep the exception if the flow failed when a client id was passed to it. We could keep the checkpoint but the status would be set to failed so it never gets run again.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt:
```diff
@@ -315,10 +315,13 @@ class TopLevelTransition(
     }
 
-    private fun retryFlowFromSafePointTransition(startingState: StateMachineState): TransitionResult {
+    private fun retryFlowFromSafePointTransition(): TransitionResult {
```
Yep, `startingState` is fed into `currentState` of the builder, so no need to also pass in here as you have figured out.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -65,13 +67,23 @@ class FlowCreator(
+            checkpoint = checkpoint,
+            fiber = fiber,
+            anyCheckpointPersisted = true,
+            reloadCheckpointAfterSuspendCount = reloadCheckpointAfterSuspendCount
```
As you pointed out this might look better with a sealed class. Do we want to do this now?

lankydan replied:
I tried it but didnt like it tbh.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -65,13 +67,23 @@ class FlowCreator(
         return createFlowFromCheckpoint(nonResidentFlow.runId, checkpoint)
```
I think you need to update this as well. Otherwise for paused flows which retry won't the count go back to zero causing it to reload the first checkpoint?

williamvigorr3 replied:
Okay I am wrong the only place we are using this is inside retryFlowFromSafePoint. This should work fine.

lankydan replied:
Yeah, pass in null from pause. It will then take the number of suspends from the checkpoint as its starting reload count (because `null` was passed to `createFlowFromCheckpoint` due to the default argument).

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -108,36 +120,51 @@ class FlowCreator(
 
     @Suppress("TooGenericExceptionCaught")
-    private inline fun <reified T : Any> tryCheckpointDeserialize(bytes: SerializedBytes<T>, flowId: StateMachineRunId): T? {
+    private inline fun <reified T : Any> tryCheckpointDeserialize(
```
Should `tryDeserializeCheckpoint` also throw a `ReloadFlowFromCheckpointException` or similar inside SingleThreadedStateMachineManager.kt? In the case that the checkpoint itself can't be deserialized.

lankydan replied:
Possibly, this is where I originally added the `ReloadFlowFromCheckpoint` exception but it turned out it wasn't being thrown from here. I could add it here to be 100% safe.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -108,36 +120,51 @@ class FlowCreator(
-    private fun Checkpoint.getFiberFromCheckpoint(runId: StateMachineRunId): FlowStateMachineImpl<*>? {
+    private fun Checkpoint.getFiberFromCheckpoint(
+        runId: StateMachineRunId,
+        reloadCheckpointAfterSuspend: Boolean
```
You could use this `@FlowCreator.reloadCheckpointAfterSuspend` (at least I think you can do this) instead of passing it in here.

lankydan replied:
Yes, I can.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -397,7 +388,7 @@ internal class SingleThreadedStateMachineManager(
             val checkpoint = tryDeserializeCheckpoint(serializedCheckpoint, flowId) ?: return
             // Resurrect flow
-            flowCreator.createFlowFromCheckpoint(flowId, checkpoint) ?: return
+            flowCreator.createFlowFromCheckpoint(flowId, checkpoint, currentState.reloadCheckpointAfterSuspendCount) ?: return
```
Maybe it would be clearer if we called an explicitly called different method here. I also think it would be better if     private val reloadCheckpointAfterSuspend = serviceHub.configuration.reloadCheckpointAfterSuspend was stored inside the statemachine manager and not inside flowCreator. The reason being is this option only actually makes a difference when called inside retryFlowFromSafePoint.

lankydan replied:
The option is also important for new flows, as it decides whether to initialise the reload count or leave it null

williamvigorr3 replied:
Actually yes this makes sense. As you set the count to zero in createCheckpointFromLogic and only increment it when a  reloadFlowFromCheckpointAfterSuspendTransition happens.

williamvigorr3 commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,429 @@
+            val handle = alice.rpc.startFlow(::ReloadFromCheckpointFlow, bob.nodeInfo.singleIdentity(), false, false, false)
+            val flowStartedByAlice = handle.id
+            handle.returnValue.getOrThrow()
+            assertEquals(null, reloadCounts[flowStartedByAlice])
```
`assertNull` :trollface:

lankydan replied:
Lol, payback for all the times I've said it to you!

williamvigorr3 commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,429 @@
+            val handle = alice.rpc.startFlow(::ReloadFromCheckpointFlow, bob.nodeInfo.singleIdentity(), false, false, false)
+            val flowStartedByAlice = handle.id
+            handle.returnValue.getOrThrow()
+            assertEquals(5, reloadCounts[flowStartedByAlice])
```
It might make it clearer to add a comments to `ReloadFromCheckpointFlow` and `ReloadFromCheckpointResponder ` every time the flow is supposed to suspend so it is easy to work out why this number should be 5.

williamvigorr3 commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,429 @@
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+class FlowReloadAfterCheckpointTest {
```
Looks comprehensive. It might be worth checking timedFlows work still by throwing a `FlowTimeoutException`.

lankydan replied:
`TimedFlow` is an `IdempotentFlow` which I have tested so it should work in the same way. Could add a test though.

lankydan replied:
Good comment. Caught a bug with retries in general and fixed it.

lankydan committed:
CORDA-3777 Keep flows observation that failed reload

lankydan committed:
CORDA-3777 Address wills pr comments

lankydan committed:
CORDA-3777 Allow normal retries

Do not let normal retries from effecting the reload count as they should
be separate.

A new event was added to handle this, so it is completely cut off from
the retry event. It still uses the retry action under the hood.

lankydan committed:
CORDA-3777 Add extra retry test

rick-r3 reviewed (commented):
Yeah, makes sense to disentangle the reload from the retry.  Happy for others to then study in more detail and approve.

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,510 @@
+
+    @Test(timeout = 300_000)
+    fun `flow will reload from its checkpoint after suspending when reloadCheckpointAfterSuspend is true`() {
+        val reloadCounts = mutableMapOf<StateMachineRunId, Int>()
```
I wonder if we should make this map a `ConcurrentMap` since the hook is accessible by 2 threads at the same time (although each mapping by one thread at a time).

lankydan replied:
possibly but there should never be contention on the same flow id.

williamvigorr3 commented:
LGTM!

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,510 @@
+    }
+
+    @Test(timeout = 300_000)
+    fun `idempotent flow will reload from initial checkpoint after calling a suspending function when reloadCheckpointAfterSuspend is true but can't throw deserialization error from objects in the call function`() {
```
This test method seems to be having the exact same body with the above one unless i m missing something.

lankydan committed:
CORDA-3777 Correct test

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,510 @@
+     */
+    @StartableByRPC
+    @InitiatingFlow
+    class MyRestartingFlow : FlowLogic<Unit>() {
```
How is this a restarting flow if it is to throw a `HospitalizeFlowException`?

lankydan replied:
Its the flow i use to hospitalise a flow so i can restart the node while the flow stays around

lankydan committed:
CORDA-3777 Correct test

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/flows/FlowReloadAfterCheckpointTest.kt:
```diff
@@ -0,0 +1,510 @@
+            )
+            val flowStartedByAlice = handle.id
+            handle.returnValue.getOrThrow(30.seconds)
+            val flowStartedByBob = bob.rpc.stateMachineRecordedTransactionMappingSnapshot()
```
I think this here could be a potential flaky situation because at this point bob's flow could have ended before testing thread executes this command.

kyriathar replied:
Actually it's transactions so i think nevermind, thought it was `stateMachinesSnapshot`.

kyriathar commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -632,8 +624,16 @@ internal class SingleThreadedStateMachineManager(
+                    "Unable to deserialize checkpoint for flow $flowId. [reloadCheckpointAfterSuspend] is turned on, throwing exception",
+                    e
+                )
+                throw ReloadFlowFromCheckpointException(e)
```
If this exception is to be used by the hospital, and in general it makes sense within the statemachine, then maybe it would make sense to check if we are inside a flow/ fiber then wrap with this exception otherwise throw the inner exception raw?

lankydan committed:
CORDA-3777 Check that the thread is a flow thread

kyriathar reviewed (dismissed):
LGTM

lankydan committed:
CORDA-3777 Use force checkpoint instead of sleep
